src/main/java/org/aksw/iguana/commons/io/BigByteArrayInputStream.java

```diff
@@ -51,71 +57,81 @@ public int read(byte[] b, int off, int len) throws IOException {
 
         if (ended) return -1;
 
-        final var copyLength1 = min(availableBytes(), len);
-        System.arraycopy(currentBuffer, posInCurrentBuffer, b, off, copyLength1);
-        posInCurrentBuffer += copyLength1;
+        final var copyLength1 = min(currentBuffer.remaining(), len);
+        currentBuffer.get(b, off, copyLength1);
         off += copyLength1;
-        if (availableBytes() == 0)
+        if (currentBuffer.remaining() == 0)
             activateNextBuffer();
 
         // check if b is already filled up or if there is nothing left to read
         if (copyLength1 == len || ended) return copyLength1;
 
-        // there might be the rare case, where reading one additional baos might not be enough to fill the buffer,
-        // because there are different array size limitations across different JVMs
-        final var copyLength2 = min(availableBytes(), len - copyLength1);
-        System.arraycopy(currentBuffer, posInCurrentBuffer, b, off, copyLength2);
-        posInCurrentBuffer += copyLength2;
+        // read again if there is still something to read
+        final var copyLength2 = min(currentBuffer.remaining(), len - copyLength1);
+        currentBuffer.get(b, off, copyLength2);
 
-        if (availableBytes() == 0)
+        if (currentBuffer.remaining() == 0)
             activateNextBuffer();
 
         return copyLength1 + copyLength2;
     }
 
+    @Override
+    public byte[] readNBytes(int n) throws IOException {
+        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
+        if (currentBuffer == null && !activateNextBuffer()) return new byte[0];
+        int read = Math.min(n, available());
+        byte[] b = new byte[read];
+        readNBytes(b, 0, read);
+        return b;
+    }
+
     @Override
     public int readNBytes(byte[] b, int off, int len) throws IOException {
         this.bbaos.close();
```

Comment:
check len against MAX_ARRAY_SIZE

Reply:
Is this necessary? Because if the given array is larger than MAX_ARRAY_SIZE, iz might not be able to exist in the first place because it's large than MAX_ARRAY_SIZE. Otherwise if the len and array size don't fit, the method will throw anyway because of `Objects.checkFromIndexSize(off, len, b.length);`.